salt/modules/freezer.py

```diff
@@ -0,0 +1,281 @@
+
+
+def _states_path():
+    '''
+    Return the path where we will store the states.
+    '''
+    return os.path.join(__opts__['cachedir'], 'freezer')
+
+
+def _paths(name=None):
+    '''
+    Return the full path for the packages and repository freezer
+    files.
+
+    '''
+    name = 'freezer' if not name else name
+    states_path = _states_path()
+    return (
+        os.path.join(states_path, '{}-pkgs.yml'.format(name)),
+        os.path.join(states_path, '{}-reps.yml'.format(name)),
+    )
+
+
+def isfrozen(name=None):
+    '''
+    Return True if there is already a frozen state.
+
+    A frozen state is merely a list of packages (including the
+    version) in a specific time. This information can be used to
+    compare with the current list of packages, and revert the
+    installation of some extra packages that are in the system.
+
+    name
+        Name of the frozen state. Optional.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' freezer.isfrozen
+
+    '''
+    name = 'freezer' if not name else name
+    return all(os.path.isfile(i) for i in _paths(name))
+
+
+def list_():
+    '''
+    Return the list of frozen states.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' freezer.list
+
+    '''
+    ret = []
+    states_path = _states_path()
+    if not os.path.isdir(states_path):
+        return ret
+
+    for state in os.listdir(states_path):
+        if state.endswith(('-pkgs.yml', '-reps.yml')):
+            # Remove the suffix, as both share the same size
+            ret.append(state[:-9])
+    return sorted(set(ret))
+
+
+def freeze(name=None, force=False, **kwargs):
+    '''
+    Save the list of package and repos in a freeze file.
+
+    As this module is build on top of the pkg module, the user can
+    send extra attributes to the underlining pkg module via kwargs.
+
+    name
+        Name of the frozen state. Optional.
+
+    force
+        If true, overwrite the state. Optional.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' freezer.freeze
+        salt '*' freezer.freeze force=True root=/chroot
+
+    '''
+    states_path = _states_path()
+    ret = __salt__['cmd.retcode'](['mkdir', '-p', states_path])
+    if ret:
+        msg = 'Error when trying to create the freezer ' \
+            'storage: {}'.format(states_path)
+        log.error(msg)
+        raise CommandExecutionError(msg)
+
+    if isfrozen(name) and not force:
+        raise CommandExecutionError('The state is already present. Use '
+                                    'force parameter to overwrite.')
+    safe_kwargs = {k: v for k, v in kwargs.items() if not k.startswith('__')}
+    pkgs = __salt__['pkg.list_pkgs'](**safe_kwargs)
+    repos = __salt__['pkg.list_repos'](**safe_kwargs)
+    for name, content in zip(_paths(name), (pkgs, repos)):
+        with fopen(name, 'w') as fp:
+            json.dump(content, fp)
+    return True
+
+
+def restore(name=None, **kwargs):
```

Comment:
We dislike allowing kwargs because a user can't know what arguments are possible. Can we either make this explicit or remove this entirely?

Reply:
I see. I have a similar comment in the `chroot` PR (https://github.com/saltstack/salt/pull/50418#issuecomment-437057166). In this case is more clear why `kwargs` is needed, as the user can send any parameter to any of the hundreds of execution modules that can be executed inside the chroot jail.

But in this case is less clear. I agree. The issue here is again _composability_. I need to compose one module, freezer, with several others that depends on. In this case `pkg.list_repos`, `pkg.mod_repo`, `pkg.list_pkgs`, `pkg.remove`, `pkg.install` and `pkg.del_repo`. The user can influence how is calling those modules via the ugly `kwargs`, for example to send the `root` parameter (my personal use case), the `refresh` to force a refresh of the repository information, or many others.

I can simply close the parameters to my single and personal use case (passing root and maybe one or two when in a week I discover that I need it), or I can do a closure of the parameters that are affected for those pkg.* modules that I use (including the code for yum, aptget, zypper, dnf, pacman and windows), and make it explicit in the signature. But this will means to have more than 15 parameters easily.

If the main problem is the user, I added in the documentation a comment about how the user can affect the underlining pkg.* modules via kwargs, but I can try to make the comment more explicit.

Reply:
I added the extra comments.

Reply:
@cachedout not sure this is quite right, we actually use `kwargs` everywhere, and we have [utility package for that](https://github.com/saltstack/salt/blob/develop/salt/utils/args.py), using it for example [here](https://github.com/saltstack/salt/blob/develop/salt/minion.py#L359-L360) or [here](https://github.com/saltstack/salt/blob/develop/salt/modules/libcloud_compute.py#L110) or [here](https://github.com/saltstack/salt/blob/develop/salt/modules/aptpkg.py#L2776) or [here](https://github.com/saltstack/salt/blob/develop/salt/modules/git.py#L3909) etc for the same reasons.

What is important is to just document that properly and raise an exception if there are unknown args, according to the examples.

Reply:
As I commented before, we actually using kwargs all over the place, and sometimes we even have to (e.g. when "async" is a parameter, but it is reserved word in Python 3.7). Still we would require:

- Proper documentation
- Validation of them with mentioned utilities

Reply:
I understand @cachedout's concerns here, but I think we just need this to be better documented. We need to tell the user what underlying functions are being called and refer the user to use documentation to determine which of them to use.